state/relationingress.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/juju/errors"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+	"gopkg.in/mgo.v2/txn"
+)
+
+// RelationIngress instances describe the ingress networks
+// required for a cross model relation.
+type RelationIngress interface {
+	RelationKey() string
+	CIDRS() []string
+}
+
+type relationIngressDoc struct {
+	Id    string   `bson:"_id"`
+	CIDRS []string `bson:"cidrs"`
+}
+
+type relationIngress struct {
+	st  *State
+	doc relationIngressDoc
+}
+
+// String returns p as a user-readable string.
+func (r *relationIngress) String() string {
+	return fmt.Sprintf("cidrs for relation %q", r.RelationKey())
+}
+
+// RelationKey is the key of the relation.
+func (r *relationIngress) RelationKey() string {
+	return r.st.localID(r.doc.Id)
+}
+
+// CIDRS returns the ingress networks for the relation.
+func (r *relationIngress) CIDRS() []string {
+	return r.doc.CIDRS
+}
+
+// RelationIngressNetworks instances provide access to relation ingress in state.
+type RelationIngressNetworks interface {
+	Save(relationKey string, cidrs []string) (RelationIngress, error)
+}
+
+type relationIngressNetworks struct {
+	st *State
+}
+
+// RelationIngressNetworks creates a RelationIngressNetworks instance backed by a state.
+func NewRelationIngressNetworks(st *State) *relationIngressNetworks {
+	return &relationIngressNetworks{st: st}
+}
+
+// Save stores the specified ingress networks for the relation.
+func (rin *relationIngressNetworks) Save(relationKey string, cidrs []string) (RelationIngress, error) {
+	logger.Debugf("save ingress networks for %v: %v", relationKey, cidrs)
+	for _, cidr := range cidrs {
+		if _, _, err := net.ParseCIDR(cidr); err != nil {
+			return nil, errors.NotValidf("CIDR %q", cidr)
+		}
+	}
+	doc := relationIngressDoc{
+		Id:    rin.st.docID(relationKey),
+		CIDRS: cidrs,
+	}
+	buildTxn := func(int) ([]txn.Op, error) {
+		model, err := rin.st.Model()
+		if err != nil {
+			return nil, errors.Annotate(err, "failed to load model")
+		}
+		if model.Life() != Alive {
```

Comment:
model.assertActiveOp() is the assertion.
There's a helper checkModelActive() which is meant to be used with the assertion so I've changed to use that instead of the straight life==alive check.